astroplan/constraints.py

```diff
@@ -394,7 +395,81 @@ def compute_constraint(self, times, observer, targets):
                              "MoonSeparationConstraint.")
         return mask
 
+    
+class DarkGreyBrightConstraint(Constraint):
+    """
+    Constrain whether it is dark, gray or bright time or how bright.
+    """
+    # TODO: make a new test
+    @u.quantity_input(horizon=u.deg)
+    def __init__(self, lunar_horizon=90*u.deg, min_moon_illumination=None,
+                                                max_moon_illumination=None):
+        '''
+        Parameters
+        ----------
+        Lunar_horizon : `~astropy.units.Quantity` (optional)
+            The altitude of the ``horizon`` and defaults to 90 degrees       
+        min_moon_illumination : float or `None` (optional)
+            the minimum value of the fraction of the moon illuminated
+        max_moon_illumination : float or `None` (optional)
+            the maximum value of the fraction of the moon illuminated
+        
+        Example
+        -------
+        Constrain the observations to times where the moon :
+
+        >>> from astroplan import Observer
+        >>> from astroplan.constraints import DarkGreyBrightConstraint
+        >>> subaru = Observer.at_site("Subaru")
+        >>> darktime = DarkGreyBrightConstraint(lunar_horizon=-5*u.deg) 
+        >>> greytime = DarkGreyBrightConstraint(max_moon_illumination=.5)
+        '''
+        
+        self.min = min_moon_illumination
+        self.max = max_moon_illumination
+        self.horizon = lunar_horizon
+        
+    def _get_lunar_altitudes(self, times, observer, targets):
+        if not hasattr(observer, '_altaz_cache'):
+            observer._altaz_cache = {}
+
+        aakey = (tuple(times.jd), 'sun')
+
+        if aakey not in observer._altaz_cache:
+            # TODO: make sure this works properly
+            # Broadcast the lunar altitudes for the number of targets:
+            altaz = observer.altaz(times, get_moon(times, observer.location))
+            altitude = altaz.alt
+            altitude.resize(1, len(altitude))
```

Comment:
I see that you've based this code on what is currently in the ```AtNightConstraint```. There are problems with that code (fixed in #165) because you're using the tuple of ```times``` as a key, but the shape of the object in the cache depends on both ```targets``` and ```times```.

Instead, I'd base this function on the ```_get_moon_data``` code in #165, and look at the ```MoonIlluminationConstraint``` in the same PR to see how to reshape the cached altitudes and illuminations to get the correct shape later on.